pipicswitch/doc/pipicswd.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PIPICSWD 1"
.TH PIPICSWD 1 "2014-02-19" "version 20140219" "Raspberry Pi"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pipicswd \-  monitor and control Raspberry Pi power switch with i2c
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpipicswd\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBpipicswd\fR is used to monitor and control Raspberry Pi power switch 
with i2c bus. The daemon is started with command
.PP
\&\fBservice\fR \fIpipicswd\fR \fIstart\fR
.PP
and can be stopped any time with
.PP
\&\fBservice\fR \fIpipicswd\fR \fIstop\fR
.PP
The operation of the daemon can be checked from the log file.
At the start four test bytes are written to the switch and read back.
If this data flow test fails the daemon exits. One reason for this
failure could be that the internal counter of \s-1PIC\s0 was not yet reset with
command \fBpipic\fR \fB\-a\fR 27 \fB\-c\fR 50.
.PP
In the main loop a socket is listened for any incoming commands. The possible
commands are
.PP
\&\fIstatus\fR send switch status to the connecting client
.PP
\&\fIclose\fR \fIN\fR close switch with channel number \fIN\fR
.PP
\&\fIopen\fR \fIN\fR open switch with channel number \fIN\fR
.SH "FILES"
.IX Header "FILES"
\&\fI/etc/logrotate.d/pipicswd\fR       Log rotation configuration file.
.PP
\&\fI/etc/init.d/pipicswd\fR            Init script.
.PP
\&\fI/etc/pipicswd_config\fR            Configuration file.
.PP
\&\fI/usr/sbin/pipicswd\fR              Daemon code.
.PP
\&\fI/var/log/pipicswd.log\fR           Log file.
.PP
\&\fI/var/run/pipicswd.pid\fR           \s-1PID\s0 file.
.SH "WARNING"
.IX Header "WARNING"
No check is done where the query data is written. Could make some hardware 
unusable.
.SH "AUTHORS"
.IX Header "AUTHORS"
Jaakko Koivuniemi
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIpipicsw\fR\|(1), \fIpipic\fR\|(1), \fIpipicfile\fR\|(1), \fIpipictest\fR\|(1), \fIsocket\fR\|(2)


pipicswitch/pipicswd.c
/**************************************************************************
 * 
 * Control external power switch using i2c bus on Raspberry Pi. The switch
 * has a PIC processor to interprete commands send by this program.  
 *       
 * Copyright (C) 2014 Jaakko Koivuniemi.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 ****************************************************************************
 *
 * Sun Feb 16 14:29:25 CET 2014
 * Edit: Mon Feb 24 21:10:13 CET 2014
 *
 * Jaakko Koivuniemi
 **/

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <unistd.h>
#include <errno.h>
#include <time.h>
#include <signal.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#define CHECK_BIT(var,pos) !!((var) & (1<<(pos)))

const int version=20140224; // program version

int portno=5001; // socket port number

float picycle=0.445; // length of PIC counter cycles [s]

const char i2cdev[100]="/dev/i2c-1";
const int  address=0x27;
const int  i2lockmax=10; // maximum number of times to try lock i2c port  

const char confile[200]="/etc/pipicswd_config";

const char pidfile[200]="/var/run/pipicswd.pid";

int loglev=3;
const char logfile[200]="/var/log/pipicswd.log";
char message[200]="";

char status[200]=""; // switch status message

int initswitch1=0; // switch 1 at start 0=do nothing, 1=switch on, 2=off
int initswitch2=0; // switch 2 at start 0=do nothing, 1=switch on, 2=off
int stopswitch1=0; // switch 1 at stop 0=do nothing, 1=switch on, 2=off
int stopswitch2=0; // switch 2 at stop 0=do nothing, 1=switch on, 2=off

void logmessage(const char logfile[200], const char message[200], int loglev, int msglev)
{
  time_t now;
  char tstr[25];
  struct tm* tm_info;
  FILE *log;

  time(&now);
  tm_info=localtime(&now);
  strftime(tstr,25,"%Y-%m-%d %H:%M:%S",tm_info);
  if(msglev>=loglev)
  {
    log=fopen(logfile, "a");
    if(NULL==log)
    {
      perror("could not open log file");
    }
    else
    { 
      fprintf(log,"%s ",tstr);
      fprintf(log,"%s\n",message);
      fclose(log);
    }
  }
}

// read configuration file if it exists
void read_config()
{
  FILE *cfile;
  char *line=NULL;
  char par[20];
  float value;
  size_t len;
  ssize_t read;

  cfile=fopen(confile, "r");
  if(NULL!=cfile)
  {
    sprintf(message,"Read configuration file");
    logmessage(logfile,message,loglev,4);

    while((read=getline(&line,&len,cfile))!=-1)
    {
       if(sscanf(line,"%s %f",par,&value)!=EOF)
       {
          if(strncmp(par,"LOGLEVEL",8)==0)
          {
             loglev=(int)value;
             sprintf(message,"Log level set to %d",(int)value);
             logmessage(logfile,message,loglev,4);
          }
          if(strncmp(par,"DCSWITCHPORT",12)==0)
          {
             portno=(int)value;
             sprintf(message,"Switch port number set to %d",(int)value);
             logmessage(logfile,message,loglev,4);
          }
          if(strncmp(par,"INITSWITCH1",11)==0)
          {
             initswitch1=(int)value;
             sprintf(message,"Switch 1 initialization set to %d",(int)value);
             logmessage(logfile,message,loglev,4);
          }
          if(strncmp(par,"INITSWITCH2",11)==0)
          {
             initswitch2=(int)value;
             sprintf(message,"Switch 2 initialization set to %d",(int)value);
             logmessage(logfile,message,loglev,4);
          }
          if(strncmp(par,"STOPSWITCH1",11)==0)
          {
             stopswitch1=(int)value;
             sprintf(message,"Switch 1 stop value set to %d",(int)value);
             logmessage(logfile,message,loglev,4);
          }
          if(strncmp(par,"STOPSWITCH2",11)==0)
          {
             stopswitch2=(int)value;
             sprintf(message,"Switch 2 stop value set to %d",(int)value);
             logmessage(logfile,message,loglev,4);
          }
          if(strncmp(par,"PICYCLE",7)==0)
          {
             picycle=value;
             sprintf(message,"PIC cycle %f s",value);
             logmessage(logfile,message,loglev,4);
          }

       }
    }
    fclose(cfile);
  }
  else
  {
    sprintf(message, "Could not open %s", confile);
    logmessage(logfile, message, loglev,4);
  }
}

// write i2c command to PIC optionally followed by data, length is the number 
// of bytes and can be 0, 1, 2 or 4 
// return: 1=ok, -1=open failed, -2=lock failed, -3=bus access failed, 
// -4=i2c slave writing failed
int write_cmd(int cmd, int data, int length)
{
  int ok=0;
  int fd,rd;
  int rnxt=0;
  int cnt=0;
  unsigned char buf[10];

  if((cmd>=0)&&(cmd<=255))
  {
    if((fd=open(i2cdev, O_RDWR)) < 0) 
    {
      sprintf(message,"Failed to open i2c port");
      logmessage(logfile,message,loglev,4);
      return -1;
    }

    rd=flock(fd, LOCK_EX|LOCK_NB);

    cnt=i2lockmax;
    while((rd==1)&&(cnt>0)) // try again if port locking failed
    {
      sleep(1);
      rd=flock(fd, LOCK_EX|LOCK_NB);
      cnt--;
    }

    if(rd)
    {
      sprintf(message,"Failed to lock i2c port");
      logmessage(logfile,message,loglev,4);
      return -2;
    }

    if(ioctl(fd, I2C_SLAVE, address) < 0) 
    {
      sprintf(message,"Unable to get bus access to talk to slave");
      logmessage(logfile,message,loglev,4);
      return -3;
    }

    buf[0]=cmd;
    if(length==1)
    { 
      buf[1]=data;
      sprintf(message,"Send 0x%02x%02x",buf[0],buf[1]);
      logmessage(logfile,message,loglev,1);
      if((write(fd, buf, 2)) != 2) 
      {
        sprintf(message,"Error writing to i2c slave");
        logmessage(logfile,message,loglev,4);
        return -4;
      }
   }
   else if(length==2)
   { 
      buf[1]=(int)(data/256);
      buf[2]=data%256;
      sprintf(message,"Send 0x%02x%02x%02x",buf[0],buf[1],buf[2]);
      logmessage(logfile,message,loglev,1); 
      if((write(fd, buf, 3)) != 3) 
      {
        sprintf(message,"Error writing to i2c slave");
        logmessage(logfile,message,loglev,4);
        return -4;
      }
   }
   else if(length==4)
   { 
      buf[1]=(int)(data/16777216);
      rnxt=data%16777216;
      buf[2]=(int)(rnxt/65536);
      rnxt=rnxt%65536;
      buf[3]=(int)(rnxt/256);
      buf[4]=rnxt%256;
      sprintf(message,"Send 0x%02x%02x%02x%02x%02x",buf[0],buf[1],buf[2],buf[3],buf[4]);
      logmessage(logfile,message,loglev,1); 
      if((write(fd, buf, 5)) != 5) 
      {
        sprintf(message,"Error writing to i2c slave");
        logmessage(logfile,message,loglev,4);
        return -4;
      }
   }
   else
   {
      sprintf(message,"Send 0x%02x",buf[0]);
      logmessage(logfile,message,loglev,1); 
      if((write(fd, buf, 1)) != 1) 
      {
        sprintf(message,"Error writing to i2c slave");
        logmessage(logfile,message,loglev,4);
        return -4;
      }
   }
   close(fd);
   ok=1;
  }
  
  return ok;
}

// read data with i2c from PIC, length is the number of bytes to read 
// return: -1=open failed, -2=lock failed, -3=bus access failed, 
// -4=i2c slave reading failed
int read_data(int length)
{
  int rdata=0;
  int fd,rd;
  int cnt=0;
  unsigned char buf[10];

  if((fd=open(i2cdev, O_RDWR)) < 0) 
  {
    sprintf(message,"Failed to open i2c port");
    logmessage(logfile,message,loglev,4);
    return -1;
  }

  rd=flock(fd, LOCK_EX|LOCK_NB);

  cnt=i2lockmax;
  while((rd==1)&&(cnt>0)) // try again if port locking failed
  {
    sleep(1);
    rd=flock(fd, LOCK_EX|LOCK_NB);
    cnt--;
  }
  if(rd)
  {
    sprintf(message,"Failed to lock i2c port");
    logmessage(logfile,message,loglev,4);
    return -2;
  }

  if(ioctl(fd, I2C_SLAVE, address) < 0) 
  {
    sprintf(message,"Unable to get bus access to talk to slave");
    logmessage(logfile,message,loglev,4);
    return -3;
  }

  if(length==1)
  {
     if(read(fd, buf,1)!=1) 
     {
       sprintf(message,"Unable to read from slave");
       logmessage(logfile,message,loglev,4);
       return -4;
     }
     else 
     {
       sprintf(message,"Receive 0x%02x",buf[0]);
       logmessage(logfile,message,loglev,1); 
       rdata=buf[0];
     }
  } 
  else if(length==2)
  {
     if(read(fd, buf,2)!=2) 
     {
       sprintf(message,"Unable to read from slave");
       logmessage(logfile,message,loglev,4);
       return -4;
     }
     else 
     {
       sprintf(message,"Receive 0x%02x%02x",buf[0],buf[1]);
       logmessage(logfile,message,loglev,1);  
       rdata=256*buf[0]+buf[1];
     }
  }
  else if(length==4)
  {
     if(read(fd, buf,4)!=4) 
     {
       sprintf(message,"Unable to read from slave");
       logmessage(logfile,message,loglev,4);
       return -4;
     }
     else 
     {
        sprintf(message,"Receive 0x%02x%02x%02x%02x",buf[0],buf[1],buf[2],buf[3]);
        logmessage(logfile,message,loglev,1);  
        rdata=16777216*buf[0]+65536*buf[1]+256*buf[2]+buf[3];
     }
  }

  close(fd);

  return rdata;
}

int testi2c()
{
  int ok=-1;
  int testint=0;
  int testres=1;

  srand((unsigned int)time(NULL));

  testint=rand();
  while((testint==-1)||(testint==-2)||(testint==-3)||(testint==-4))
  {
    testint=rand();
  }

  ok=write_cmd(0x02,testint,4);

  if(ok!=1)
  {
    strcpy(message,"failed to write 4 test bytes"); 
    logmessage(logfile,message,loglev,4);
  }
  else
  {
    testres=read_data(4);

    if((testres==-1)||(testres==-2)||(testres==-3)||(testres==-4))
    {
      strcpy(message,"failed to read 4 test bytes"); 
      logmessage(logfile,message,loglev,4);
    }
    else
    {
      if(testint==testres) ok=1; else ok=0;
    }
  }

  return ok;
}

// read switch status
int read_status()
{
  int ok=-1;
  int gpio=0;

  ok=write_cmd(0x01,0x05,1); 
  if(ok==1)
  { 
    gpio=read_data(1);
    if(CHECK_BIT(gpio,4)==1) 
    {
      strcpy(message,"switch 1 closed");
      strcpy(status,"1 closed"); 
    }
    else 
    {
      strcpy(message,"switch 1 open");
      strcpy(status,"1 open");
    }

    if(CHECK_BIT(gpio,5)==1) 
    {
      strcat(message," switch 2 closed");
      strcat(status," 2 closed");
    }
    else 
    {
      strcat(message," switch 2 open");
      strcat(status," 2 open");
    }
    logmessage(logfile,message,loglev,3);
  }
  else
  {
    strcpy(message,"failed to read PIC GPIO register"); 
    logmessage(logfile,message,loglev,4);
  }

  return ok;
}

int operate_switch1(int switch1)
{
  int ok=0;

  if(switch1==1) 
  {
     ok=write_cmd(0x24,0x00,0); 
     if(ok!=1) 
     {
       sprintf(message,"closing switch 1 failed");
       logmessage(logfile,message,loglev,4);
     }
     else
     {
       sprintf(message,"switch 1 closed");
       logmessage(logfile,message,loglev,2);
     }

  }
  else if(switch1==2) 
  {
     ok=write_cmd(0x14,0x00,0); 
     if(ok!=1) 
     {
       sprintf(message,"opening switch 1 failed");
       logmessage(logfile,message,loglev,4);
     }
     else
     {
       sprintf(message,"switch 1 opened");
       logmessage(logfile,message,loglev,2);
     }
  }

  return ok;
}

int operate_switch2(int switch2)
{
  int ok=0;

  if(switch2==1) 
  {
     ok=write_cmd(0x25,0x00,0); 
     if(ok!=1) 
     {
       sprintf(message,"closing switch 2 failed");
       logmessage(logfile,message,loglev,4);
     }
     else
     {
       sprintf(message,"switch 2 closed");
       logmessage(logfile,message,loglev,2);
     }
  }
  else if(switch2==2) 
  {
     ok=write_cmd(0x15,0x00,0); 
     if(ok!=1) 
     {
       sprintf(message,"opening switch 2 failed");
       logmessage(logfile,message,loglev,4);
     }
     else
     {
       sprintf(message,"switch 2 opened");
       logmessage(logfile,message,loglev,2);
     }
  }

  return ok;
}



int cont=1; /* main loop flag */

void stop(int sig)
{
  sprintf(message,"signal %d catched, stop",sig);
  logmessage(logfile,message,loglev,4);
  cont=0;
}

void terminate(int sig)
{
  int ok=0;

  sprintf(message,"signal %d catched",sig);
  logmessage(logfile,message,loglev,4);

  sleep(1);
  ok=operate_switch1(stopswitch1);
  ok=operate_switch2(stopswitch2);

  sleep(1);
  strcpy(message,"stop");
  logmessage(logfile,message,loglev,4);

  cont=0;
}

void hup(int sig)
{
  sprintf(message,"signal %d catched",sig);
  logmessage(logfile,message,loglev,4);
}



int main()
{  
  int ok=0;

  sprintf(message,"pipicswd v. %d started",version); 
  logmessage(logfile,message,loglev,4);

  signal(SIGINT,&stop); 
  signal(SIGKILL,&stop); 
  signal(SIGTERM,&terminate); 
  signal(SIGQUIT,&stop); 
  signal(SIGHUP,&hup); 

  int i2cok=testi2c(); // test i2c data flow to PIC 
  if(i2cok==1)
  {
    strcpy(message,"i2c dataflow test ok"); 
    logmessage(logfile,message,loglev,4);
  }
  else
  {
    strcpy(message,"i2c dataflow test failed, exit");
    logmessage(logfile,message,loglev,4); 
    strcpy(message,"try to reset timer with 'pipic -a 27 -c 50'");
    logmessage(logfile,message,loglev,4); 
    strcpy(message,"and restart with 'service pipicswd start'");
    logmessage(logfile,message,loglev,4); 
    cont=0;
  }

  read_config(); // read configuration file

  pid_t pid, sid;
        
  pid=fork();
  if(pid<0) 
  {
    exit(EXIT_FAILURE);
  }

  if(pid>0) 
  {
    exit(EXIT_SUCCESS);
  }

  umask(0);

  /* Create a new SID for the child process */
  sid=setsid();
  if(sid<0) 
  {
    strcpy(message,"failed to create child process"); 
    logmessage(logfile,message,loglev,4);
    exit(EXIT_FAILURE);
  }
        
  if((chdir("/")) < 0) 
  {
    strcpy(message,"failed to change to root directory"); 
    logmessage(logfile,message,loglev,4);
    exit(EXIT_FAILURE);
  }
        
  /* Close out the standard file descriptors */
  close(STDIN_FILENO);
  close(STDOUT_FILENO);
  close(STDERR_FILENO);

  FILE *pidf;
  pidf=fopen(pidfile,"w");

  if(pidf==NULL)
  {
    sprintf(message,"Could not open PID lock file %s, exiting", pidfile);
    logmessage(logfile,message,loglev,4);
    exit(EXIT_FAILURE);
  }

  if(flock(fileno(pidf),LOCK_EX||LOCK_NB)==-1)
  {
    sprintf(message,"Could not lock PID lock file %s, exiting", pidfile);
    logmessage(logfile,message,loglev,4);
    exit(EXIT_FAILURE);
  }

  fprintf(pidf,"%d\n",getpid());
  fclose(pidf);

// open socket
  int sockfd, connfd, clilen;
  char rbuff[25];
  char sbuff[25];
  struct sockaddr_in serv_addr, cli_addr; 

  sockfd=socket(AF_INET, SOCK_STREAM, 0);
  if(sockfd<0) 
  {
    sprintf(message,"Could not open socket");
    logmessage(logfile,message,loglev,4);
    exit(EXIT_FAILURE);
  }
  else
  {
    sprintf(message,"Socket open");
    logmessage(logfile,message,loglev,2);
  }
  
  memset(&serv_addr, '0', sizeof(serv_addr));
  memset(sbuff, '0', sizeof(sbuff)); 

  serv_addr.sin_family = AF_INET;
  serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
  serv_addr.sin_port = htons(portno); 

  if(bind(sockfd, (struct sockaddr*)&serv_addr, sizeof(serv_addr))<0)
  {
    sprintf(message,"Could not bind socket");
    logmessage(logfile,message,loglev,4);
    exit(EXIT_FAILURE);
  }
  else
  {
    sprintf(message,"Socket binding successful");
    logmessage(logfile,message,loglev,2);
  }

  listen(sockfd, 1); // listen one client only 
  clilen=sizeof(cli_addr);

// initialize switches
  ok=operate_switch1(initswitch1);
  ok=operate_switch2(initswitch2);
 
  sleep(1);
  ok=read_status();

  int n=0;
  while(cont==1)
  {
    connfd=accept(sockfd, (struct sockaddr*)&cli_addr, (socklen_t *)&clilen); 
    if(connfd<0) 
    {
      sprintf(message,"Socket accept failed");
      logmessage(logfile,message,loglev,4);
      exit(EXIT_FAILURE);
    }
    else
    {
      sprintf(message,"Socket accepted");
      logmessage(logfile,message,loglev,2);
    }

    bzero(rbuff,25);
    n=read(connfd,rbuff,24);
    if(n<0)
    {
      sprintf(message,"Socket reading failed");
      logmessage(logfile,message,loglev,4);
      exit(EXIT_FAILURE);
    }
    else
    {
      sprintf(message,"Received: %s",rbuff);
      logmessage(logfile,message,loglev,3);
    }

    if(strncmp(rbuff,"open 1",6)==0)
    {
      ok=operate_switch1(2);
      sleep(1);
    } 
    else if(strncmp(rbuff,"close 1",7)==0)
    {
      ok=operate_switch1(1);
      sleep(1);
    } 
    else if(strncmp(rbuff,"open 2",6)==0)
    {
      ok=operate_switch2(2);
      sleep(1);
    } 
    else if(strncmp(rbuff,"close 2",7)==0)
    {
      ok=operate_switch2(1);
      sleep(1);
    } 

    ok=read_status();
    snprintf(sbuff,sizeof(sbuff),"%.24s",status);

    n=write(connfd,sbuff,strlen(sbuff)); 
    if(n<0)
    {
      sprintf(message,"Socket writing failed");
      exit(EXIT_FAILURE);
    }
    else
    {
      sprintf(message,"Send: %s",sbuff);
      logmessage(logfile,message,loglev,3);
    }

    close(connfd);

    sleep(1);
  }

  strcpy(message,"remove PID file");
  logmessage(logfile,message,loglev,4);
  ok=remove(pidfile);

  return ok;
}
